Feat: OPA overview + changelog (#481)
Kong/developer.konghq.com: app/_kong_plugins/opa/index.md

search_aliases:
  - open policy agent

related_resources:
  - text: Block unauthorized requests in {{site.base_gateway}} with the OPA plugin
    url: /how-to/block-unauthorized-requests-with-opa/
---

## Overview

The OPA plugin allows you to forward requests to [Open Policy Agent](https://openpolicyagent.org/) and process the requests only if the authorization policy allows it.

## OPA input

> When {{site.base_gateway}} forwards a request to OPA, the request data is sent with the following structure:

Reviewer: ```suggestion
Depending on what you configure for the plugin, the following request data is sent when {{site.base_gateway}} forwards a request to OPA:
```

Author: The request data sent doesn't depend on the plugin config, it depends on the request sent. And "the following request data is sent" makes it sound like the whole structure is sent, which isn't the case, it depends on the contents of the request. i'll rephrase this differently.

     "http": {
       "host": "example.org", # host header used by the client to make the request
       "port": "8000",        # port to which the request was made
       "tls": {},             # TLS details if the request was made on HTTPS and Kong terminated the TLS connection
       "method": "GET",       # HTTP method used in the request
       "scheme": "http",      # protocol used to make the request by the client, this can be either `http` or `https`
       "path": "/foo/bar",    # HTTP path used in the request
       "querystring": {       # Query string in the HTTP request as key-value pairs
         "foo" : "bar",
         "foo2" : "bar2",
       },
       "headers": {           # HTTP headers of the request
         "accept-encoding": "gzip, deflate",
         "connection": "keep-alive",
         "accept": "*\\/*"
       },
       "uri_captures": {      # The regex capture groups captured on the Kong Gateway Route's path field in the current request. Injected only if `include_uri_captures_in_opa_input` is set to `true`.
         "named": {},
         "unnamed": []
       }
     }
   },
   "client_ip": "127.0.0.1",# client IP address as interpreted by Kong

>    "service": {             # The Kong service resource to which this request will be forwarded to if OPA allows. Injected only if `include_service_in_opa_input` is set to `true`.

Reviewer: ```suggestion
   "service": {             # The Kong service resource that this request is forwarded to if OPA allows. Injected only if `include_service_in_opa_input` is set to `true`.
```

     "ws_id": "d6020dc4-67f5-4c62-8b45-e2f497c20f5c",
     "request_buffering": true,
     "updated_at": 1612819949,
     "preserve_host": false,
     "regex_priority": 0,
     "response_buffering": true,
     "https_redirect_status_code": 426,
     "path_handling": "v0",
     "service": {
       "id": "e6fd8b19-89e5-44e6-8a2a-79e8bf3c31a5"
     }
   },
   "consumer": {            # Kong consumer that was used for authentication for this request. Injected only if `include_consumer_in_opa_input` is set to `true`.
     "id": "bc6d8617-76a7-431f-aa40-32eb1f5be7e6",
     "username": "kong-consumer-username"
   }
 }
}
```

You can use this structure, along with the [OPA docs](https://www.openpolicyagent.org/docs/latest/policy-language/), as a reference to create your policies. For example, you can use `input.request.http.querystring.page` to refer to a `page` query parameter.

## OPA response

Once OPA is done executing policies, the plugin expects the policy evaluation result as either a boolean or an object. If OPA returns any other format or a status code other than `200 OK`, the plugin will return a `500 Internal Server Error` to the client.

### Boolean response

> OPA can return a `true` or `false` result after a policy evaluation. If the input request meets the defined policies, OPA should send the following response:
> ```json
> {
>  "result": true
> }
> ```
> 
> If the request violates the policy, OPA should send the following response:
> ```json
> {
>  "result": false
> }
> ```
> In this case, any other fields in the response are ignored.

Reviewer: ```suggestion
OPA can return a `true` or `false` result after a policy evaluation. If the input request meets the defined policies, OPA should send a `"result": true` response.  If the request violates the policy, OPA should send a `"result": false` response. In this case, any other fields in the response are ignored.
```
Shortening this since there's not much in the JSON example.
